conans/test/functional/toolchains/cmake/test_cmake.py

```diff
@@ -528,3 +528,51 @@ def build(self):
         client.run("install .")
         client.run("build .")
         self.assertIn("VALUE OF CONFIG STRING: my new value", client.out)
+
+
+@pytest.mark.skipif(platform.system() != "Windows", reason="Tests Windows MinGW")
+class TestMinGW:
+    conanfile = textwrap.dedent("""
+            from conans import ConanFile
+            from conan.tools.cmake import CMake, CMakeToolchain
+            class App(ConanFile):
+                settings = "os", "arch", "compiler", "build_type"
+                generators = "cmake_find_package_multi"
+                exports_sources = "CMakeLists.txt", "main.cpp"
+
+                def generate(self):
+                    tc = CMakeToolchain(self)
+                    tc.generate()
+
+                def build(self):
+                    cmake = CMake(self)
+                    cmake.configure()
+                    cmake.build()
+            """)
+    cmakelists = textwrap.dedent("""
+        cmake_minimum_required(VERSION 2.8)
+        project(app)
+        add_executable(app main.cpp)
+        """)
+    main_cpp = gen_function_cpp(name="main")
+
+    @pytest.mark.tool_mingw64
+    @pytest.mark.tool_cmake(version="3.15")
+    def test_mingw64(self):
+        profile = textwrap.dedent("""
+            [settings]
+            os=Windows
+            arch=x86_64
+            build_type=Release
+            compiler=gcc
+            compiler.version=4.9
+            compiler.libcxx=libstdc++
+            compiler.cppstd=98
+            """)
+        client = TestClient()
+        client.save({"conanfile.py": self.conanfile, "CMakeLists.txt": self.cmakelists,
+                     "main.cpp": self.main_cpp, "profile": profile})
+        client.run_command("cmake --version")
+        assert "cmake version 3.15" in client.out
+        client.run_command('cmake . -G "MinGW Makefiles" -DCMAKE_SH=CMAKE_SH-NOTFOUND')
```

Comment:
I have investigated, and yes, it is hardcoded in the CMake codebase:

```cpp
if (this->AlreadyInCache) {
    // If the user specifies the entry on the command line without a
    // type we should add the type and docstring but keep the original
    // value.
    if (this->AlreadyInCacheWithoutMetaInfo) {
      this->Makefile->AddCacheDefinition(this->VariableName, "",
                                         this->VariableDocumentation.c_str(),
                                         cmStateEnums::FILEPATH);
    }
    return true;
  }
```
This is the only way it will work, in the command line argument. Please add it to the command line conditionally to MinGW and Windows